pkg/host/host.go

```diff
@@ -0,0 +1,95 @@
+package host
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+	"github.com/Mellanox/nic-configuration-operator/pkg/consts"
+)
+import "github.com/jaypipes/ghw"
+
+type HostManager interface {
+	DiscoverNicDevices() map[string]v1alpha1.NicDeviceStatus
+	UpdateDeviceConditionsStatus(device *v1alpha1.NicDeviceStatus)
+}
+
+type hostManager struct {
+	nodeName  string
+	hostUtils HostUtils
+}
+
+func (h hostManager) DiscoverNicDevices() map[string]v1alpha1.NicDeviceStatus {
+	log.Log.Info("HostManager.DiscoverNicDevices()")
+
+	pci, err := ghw.PCI()
+	if err != nil {
+		log.Log.Error(err, "Failed to read PCI devices")
+		return nil
+	}
+
+	// Map of Serial Number to nic device
+	devices := make(map[string]v1alpha1.NicDeviceStatus)
+
+	for _, device := range pci.Devices {
+		if device.Vendor.ID == consts.MELLANOX_VENDOR {
+			log.Log.Info("Found Mellanox device", "address", device.Address, "type", device.Product.Name)
+
+			if h.hostUtils.IsSriovVF(device.Address) {
+				log.Log.Info("Device is an SRIOV VF, skipping", "address", device.Address)
+			}
+
+			partNumber, serialNumber, err := h.hostUtils.GetPartAndSerialNumber(device.Address)
+			if err != nil {
+				log.Log.Error(err, "Failed to process device, skipping", "address", device.Address)
+				continue
+			}
+
+			// Devices with the same serial number are ports of the same NIC, so grouping them
+			deviceStatus, ok := devices[serialNumber]
+
+			if !ok {
+				firmwareVersion, psid, err := h.hostUtils.GetFirmwareVersionAndPSID(device.Address)
+				if err != nil {
+					log.Log.Error(err, "Failed to process device, skipping", "address", device.Address)
+					continue
+				}
+
+				deviceStatus = v1alpha1.NicDeviceStatus{
+					Node:            h.nodeName,
+					Type:            device.Product.ID,
+					SerialNumber:    serialNumber,
+					PartNumber:      partNumber,
+					PSID:            psid,
+					FirmwareVersion: firmwareVersion,
+					Ports:           []v1alpha1.NicDevicePortSpec{},
+				}
+
+				devices[serialNumber] = deviceStatus
+			}
+
+			networkInterface := h.hostUtils.GetInterfaceName(device.Address)
+			rdmaInterface := h.hostUtils.GetRDMADeviceName(device.Address)
+
+			deviceStatus.Ports = append(deviceStatus.Ports, v1alpha1.NicDevicePortSpec{
+				PCI:              device.Address,
+				NetworkInterface: networkInterface,
+				RdmaInterface:    rdmaInterface,
+			})
+
+			devices[deviceStatus.SerialNumber] = deviceStatus
+		}
+	}
+
+	return devices
+}
+
+func (h hostManager) UpdateDeviceConditionsStatus(device *v1alpha1.NicDeviceStatus) {
+	device.Conditions = []v1alpha1.NicDeviceConditionSpec{{
```

Comment:
there is a lib in api machinery to deal with conditions. i strongly suggest to use it.

as an example. you must set last transition time, which it will take care for u.
moreover if there are other conditions this function will remove them and replace with this single condition.

Reply:
Done, the logic will evolve, now added only a basic condition